src/Extensions/Cabal.hs

```diff
@@ -27,25 +32,50 @@ import Distribution.Types.GenericPackageDescription (GenericPackageDescription (
 import Distribution.Types.Library (Library (..))
 import Distribution.Types.TestSuite (TestSuite (..))
 import Distribution.Types.TestSuiteInterface (TestSuiteInterface (..))
-import Distribution.Verbosity (normal)
 import GHC.LanguageExtensions.Type (Extension (..))
 import System.Directory (doesFileExist)
 import System.FilePath ((<.>), (</>))
 
 import Extensions.OnOff (OnOffExtension (..))
 
+import qualified Data.ByteString as ByteString
 import qualified Data.Map.Strict as Map
 import qualified Language.Haskell.Extension as Cabal
 
 
+{- | Exception that gets thrown when working with @.cabal@ files.
+-}
+data CabalException
+    -- | The @.cabal@ file is not found.
+    = CabalFileNotFound FilePath
+    -- | Parsing errors in the @.cabal@ file.
+    | CabalParseError
+    deriving stock (Show, Eq)
+    deriving anyclass (Exception)
 
 {- | Parse default extensions from a @.cabal@ file under given
 'FilePath'.
+
+__Throws__:
+
+* 'CabalException'
+-}
+parseCabalFileExtensions :: FilePath -> IO (Map FilePath [OnOffExtension])
+parseCabalFileExtensions cabalPath = doesFileExist cabalPath >>= \hasCabalFile ->
+    if hasCabalFile
+    then ByteString.readFile cabalPath >>= parseCabalExtensions
+    else throwIO $ CabalFileNotFound cabalPath
+
+{- | Parse default extensions from a @.cabal@ file content.
+
+__Throws__:
+
+* 'CabalException'
 -}
-parseCabalExtensions :: FilePath -> IO (Map FilePath [OnOffExtension])
-parseCabalExtensions cabalPath = do
-    pkgDesc <- readGenericPackageDescription normal cabalPath
-    extractCabalExtensions pkgDesc
+parseCabalExtensions :: ByteString -> IO (Map FilePath [OnOffExtension])
+parseCabalExtensions cabal = case parseGenericPackageDescriptionMaybe cabal of
+    Nothing      -> throwIO CabalParseError
```

Comment:
This is a bit tricky part, since this function can throw only `CabalParseError` constructor, but I'm not sure what is better: one exception type with two constructors, two exception types or some type-level magic.

Reply:
I think this is enough at the moment. Probably could be improved later, but I don't see how this could be a stopper 🙂